### DISCUSSION TITLE
How does libfuse ensure that fuse init will be processed first?

### DISCUSSION THREAD

**Discussion by JiaoDean** [libfuse/libfuse#877] [General]
We modified libfuse and the kernel. When the backend hangs, the fuse_init request will be resent, and then the req stuck in the kernel will be sent. 

There has always been a doubt about how libfuse and the kernel ensure that fuse_init is processed before other fuse_type requests. 

Can you show me the code?

**Comment by Nikratio** [comment]
I don't think there's anything in libfuse to ensure that. We depend on the kernel to not send requests until the init response has been received.

I haven't checked the code though.

**Comment by amitgeron** [comment]
Have a look over here:
https://github.com/libfuse/libfuse/blob/bd8985945b5a1c35b7baa71571b1b61807ad7514/lib/fuse_lowlevel.c#L2712

**Reply by JiaoDean** [reply]
"got_init" is the logic of libfuse. The initialization process involves the 'mount' command -> fuse_init -> libfuse got_init. What I would like to ask is, in the kernel, how can the fuse_init request be sent to libfuse before other requests? Although it's not the answer I was looking for, I still want to thank you for your enthusiasm. Thank you.

**Comment by bsbernd** [comment]
Kernel protection is:

fuse_get_req()
```

	if (fuse_block_alloc(fc, for_background)) {
		err = -EINTR;
		if (wait_event_killable_exclusive(fc->blocked_waitq,
				!fuse_block_alloc(fc, for_background)))
			goto out;
	}
```

```
static bool fuse_block_alloc(struct fuse_conn *fc, bool for_background)
{
	return !fc->initialized || (for_background && fc->blocked);
}
```

process_init_reply()
    fuse_set_initialized()
```
void fuse_set_initialized(struct fuse_conn *fc)
{
	/* Make sure stores before this are seen on another CPU */
	smp_wmb();
	fc->initialized = 1;
}
```

**Reply by JiaoDean** [reply]
Thank you for your answer. It seems that this part has resolved my doubts.
